lib/streamlit/server/Server.py

```diff
@@ -429,52 +428,64 @@ def add_preheated_report_session(self):
-            The newly-connected websocket handler
+        ws : _BrowserWebSocketHandler or PREHEATED_REPORT_SESSION
+            The newly-connected websocket handler.
 
         Returns
         -------
         ReportSession
-            The ReportSession associated with this browser connection
+            The newly-created ReportSession for this browser connection.
 
         """
-        if ws not in self._session_infos:
+        if PREHEATED_REPORT_SESSION in self._session_info_by_id:
+            assert len(self._session_info_by_id) == 1
+            LOGGER.debug("Reusing preheated context for ws %s", ws)
+            session = self._session_info_by_id[PREHEATED_REPORT_SESSION].session
+            del self._session_info_by_id[PREHEATED_REPORT_SESSION]
+        else:
+            LOGGER.debug("Creating new context for ws %s", ws)
+            session = ReportSession(
+                ioloop=self._ioloop,
+                script_path=self._script_path,
+                command_line=self._command_line,
+            )
 
-            if PREHEATED_REPORT_SESSION in self._session_infos:
-                assert len(self._session_infos) == 1
-                LOGGER.debug("Reusing preheated context for ws %s", ws)
-                session = self._session_infos[PREHEATED_REPORT_SESSION].session
-                del self._session_infos[PREHEATED_REPORT_SESSION]
-            else:
-                LOGGER.debug("Creating new context for ws %s", ws)
-                session = ReportSession(
-                    ioloop=self._ioloop,
-                    script_path=self._script_path,
-                    command_line=self._command_line,
-                )
+        assert session.id not in self._session_info_by_id, (
+            "session.id '%s' registered multiple times!" % session.id
+        )
+        self._session_info_by_id[session.id] = SessionInfo(ws, session)
+
+        if ws is not PREHEATED_REPORT_SESSION:
+            self._set_state(State.ONE_OR_MORE_BROWSERS_CONNECTED)
+
+        return session
 
-            self._session_infos[ws] = SessionInfo(session)
+    def _close_report_session(self, session_id):
+        """Shutdown and remove a ReportSession.
 
-            if ws is not PREHEATED_REPORT_SESSION:
-                self._set_state(State.ONE_OR_MORE_BROWSERS_CONNECTED)
+        This function may be called multiple times for the same session,
+        which is not an error. (Subsequent calls just no-op.)
 
-        return self._session_infos[ws].session
+        Parameters
+        ----------
+        session_id : str
+            The ReportSession's id string.
 
-    def _remove_browser_connection(self, ws):
-        if ws in self._session_infos:
-            session_info = self._session_infos[ws]
-            del self._session_infos[ws]
+        """
+        if session_id in self._session_info_by_id:
+            session_info = self._session_info_by_id[session_id]
+            del self._session_info_by_id[session_id]
```

Comment:
Another function? BTW, fine if you think this is just a couple of extra lines and it is more readable.